From afa38ff2b80bcbee8d407eae7aa0e957c90c86cd Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.tech>
Date: Thu, 24 Jul 2025 16:01:33 +0200
Subject: [PATCH 2/3] Boot on Alma10

* xen.efi cannot yet be booted using multiboot2, stay with xen.gz
* use full name for vmlinuz and initrd, not the XS custom symlink
* explicitly load multiboot2 (not sure why any insmod here would work
  any better than from the GRUB shell, with prefix being wrongly set
---
 backend.py | 31 +++++++++++++++++--------------
 1 file changed, 17 insertions(+), 14 deletions(-)

diff --git a/backend.py b/backend.py
index 39b116b..e942542 100644
--- a/backend.py
+++ b/backend.py
@@ -966,10 +966,10 @@ def mkinitrd(mounts, primary_disk, primary_partnum):
 def prepFallback(mounts, primary_disk, primary_partnum):
     kernel_version =  getKernelVersion(mounts['root'])
 
-    # Copy /boot/xen-xxxx.efi to /boot/xen-fallback.efi
-    xen_efi = os.path.realpath(mounts['root'] + "/boot/xen.efi")
+    # Copy /boot/xen-xxxx.gz to /boot/xen-fallback.gz
+    xen_efi = os.path.realpath(mounts['root'] + "/boot/xen.gz")
     src = os.path.join(mounts['root'], "boot", os.path.basename(xen_efi))
-    dst = os.path.join(mounts['root'], 'boot/xen-fallback.efi')
+    dst = os.path.join(mounts['root'], 'boot/xen-fallback.gz')
     shutil.copyfile(src, dst)
 
     # Copy /boot/vmlinuz-yyyy to /boot/vmlinuz-fallback
@@ -1026,31 +1026,34 @@ def buildBootLoaderMenu(mounts, xen_version, xen_kernel_version, boot_config, se
         for interface in fcoe_interfaces:
             common_kernel_params += " fcoe=%s:%s" % (netutil.getHWAddr(interface), 'nodcb' if fcoeutil.hw_lldp_capable(interface) else 'dcb')
 
-    e = bootloader.MenuEntry(hypervisor="/boot/xen.efi",
+    e = bootloader.MenuEntry(hypervisor="/boot/xen.gz",
                              hypervisor_args=' '.join([common_xen_params, common_xen_unsafe_params, xen_mem_params, "console=vga vga=mode-0x0311"]),
-                             kernel="/boot/vmlinuz-%s-xen" % short_version,
+                             kernel=f"/boot/vmlinuz-{xen_kernel_version}",
                              kernel_args=' '.join([common_kernel_params, kernel_console_params, "console=tty0 quiet vga=785 splash plymouth.ignore-serial-consoles"]),
-                             initrd="/boot/initrd-%s-xen.img" % short_version, title=MY_PRODUCT_BRAND,
+                             initrd=f"/boot/initrd-{xen_kernel_version}.img",
+                             title=MY_PRODUCT_BRAND,
                              root=constants.rootfs_label%disk_label_suffix)
     boot_config.append("xe", e)
     boot_config.default = "xe"
     if serial:
         xen_serial_params = "%s console=%s,vga" % (serial.xenFmt(), serial.port)
 
-        e = bootloader.MenuEntry(hypervisor="/boot/xen.efi",
+        e = bootloader.MenuEntry(hypervisor="/boot/xen.gz",
                                  hypervisor_args=' '.join([xen_serial_params, common_xen_params, common_xen_unsafe_params, xen_mem_params]),
-                                 kernel="/boot/vmlinuz-%s-xen" % short_version,
+                                 kernel=f"/boot/vmlinuz-{xen_kernel_version}",
                                  kernel_args=' '.join([common_kernel_params, "console=tty0", kernel_console_params]),
-                                 initrd="/boot/initrd-%s-xen.img" % short_version, title=MY_PRODUCT_BRAND+" (Serial)",
+                                 initrd=f"/boot/initrd-{xen_kernel_version}.img",
+                                 title=MY_PRODUCT_BRAND+" (Serial)",
                                  root=constants.rootfs_label%disk_label_suffix)
         boot_config.append("xe-serial", e)
         if boot_serial:
             boot_config.default = "xe-serial"
-        e = bootloader.MenuEntry(hypervisor="/boot/xen.efi",
+        e = bootloader.MenuEntry(hypervisor="/boot/xen.gz",
                                  hypervisor_args=' '.join([safe_xen_params, common_xen_params, xen_serial_params]),
-                                 kernel="/boot/vmlinuz-%s-xen" % short_version,
+                                 kernel=f"/boot/vmlinuz-{xen_kernel_version}",
                                  kernel_args=' '.join(["earlyprintk=xen", common_kernel_params, "console=tty0", kernel_console_params]),
-                                 initrd="/boot/initrd-%s-xen.img" % short_version, title=MY_PRODUCT_BRAND+" in Safe Mode",
+                                 initrd=f"/boot/initrd-{xen_kernel_version}.img",
+                                 title=MY_PRODUCT_BRAND+" in Safe Mode",
                                  root=constants.rootfs_label%disk_label_suffix)
         boot_config.append("safe", e)
 
@@ -1060,7 +1063,7 @@ def buildBootLoaderMenu(mounts, xen_version, xen_kernel_version, boot_config, se
                             root=constants.rootfs_label%disk_label_suffix)
     boot_config.append("memtest", e)
 
-    e = bootloader.MenuEntry(hypervisor="/boot/xen-fallback.efi",
+    e = bootloader.MenuEntry(hypervisor="/boot/xen-fallback.gz",
                              hypervisor_args=' '.join([common_xen_params, common_xen_unsafe_params, xen_mem_params]),
                              kernel="/boot/vmlinuz-fallback",
                              kernel_args=' '.join([common_kernel_params, kernel_console_params, "console=tty0"]),
@@ -1069,7 +1072,7 @@ def buildBootLoaderMenu(mounts, xen_version, xen_kernel_version, boot_config, se
                              root=constants.rootfs_label%disk_label_suffix)
     boot_config.append("fallback", e)
     if serial:
-        e = bootloader.MenuEntry(hypervisor="/boot/xen-fallback.efi",
+        e = bootloader.MenuEntry(hypervisor="/boot/xen-fallback.gz",
                                  hypervisor_args=' '.join([xen_serial_params, common_xen_params, common_xen_unsafe_params, xen_mem_params]),
                                  kernel="/boot/vmlinuz-fallback",
                                  kernel_args=' '.join([common_kernel_params, "console=tty0", kernel_console_params]),
-- 
2.39.5

